### PR TITLE
TSDB: Reject the nested object fields that are configured time_series_dimension

### LINKED ISSUE #83915: tsdb/05_dimension_and_metric_in_non_tsdb_index/nested dimensions
### CI Link

local

### Repro line

 ./gradlew ':rest-api-spec:yamlRestTest' --tests "org.elasticsearch.test.rest.ClientYamlTestSuiteIT.test {yaml=tsdb/05_dimension_and_metric_in_non_tsdb_index/nested dimensions}" -Dtests.seed=CE43030603A1AB90 -Dtests.locale=en-PH -Dtests.timezone=Asia/Aqtau -Druntime.java=17

### Does it reproduce?

Yes

### Applicable branches

master

### Failure history

I just caused it

### Failure excerpt
```
 java.lang.AssertionError: Failure at [tsdb/05_dimension_and_metric_in_non_tsdb_index:239]: expected [2xx] status code but api [index] returned [400 Bad Request] [{"error":{"root_cause":[{"type":"mapper_parsing_exception","reason":"failed to parse field [nested.dim] of type [keyword] in document with id 'yF2M-n4BxnJzvZeHLUG1'. Preview of field's value: 'bar'","stack_trace":"org.elasticsearch.index.mapper.MapperParsingException: failed to parse field [nested.dim] of type [keyword] in document with id 'yF2M-n4BxnJzvZeHLUG1'. Preview of field's value: 'bar'\n\tat org.elasticsearch.index.mapper.FieldMapper.parse(FieldMapper.java:254)\n\tat org.elasticsearch.index.mapper.DocumentParser.parseObjectOrField(DocumentParser.java:356)\n\tat org.elasticsearch.index.mapper.DocumentParser.parseValue(DocumentParser.java:508)\n\tat org.elasticsearch.index.mapper.DocumentParser.innerParseObject(DocumentParser.java:296)\n\tat org.elasticsearch.index.mapper.DocumentParser.parseObjectOrNested(DocumentParser.java:260)\n\tat org.elasticsearch.index.mapper.DocumentParser.parseObjectOrField(DocumentParser.java:354)\n\tat org.elasticsearch.index.mapper.DocumentParser.parseObject(DocumentParser.java:387)\n\tat org.elasticsearch.index.mapper.DocumentParser.parseNonDynamicArray(DocumentParser.java:470)\n\tat org.elasticsearch.index.mapper.DocumentParser.parseArray(DocumentParser.java:428)\n\tat org.elasticsearch.index.mapper.DocumentParser.innerParseObject(DocumentParser.java:284)\n\tat org.elasticsearch.index.mapper.DocumentParser.parseObjectOrNested(DocumentParser.java:260)\n\tat org.elasticsearch.index.mapper.DocumentParser.internalParseDocument(DocumentParser.java:118)\n\tat org.elasticsearch.index.mapper.DocumentParser.parseDocument(DocumentParser.java:80)\n\tat org.elasticsearch.index.mapper.DocumentMapper.parse(DocumentMapper.java:80)\n\tat org.elasticsearch.index.shard.IndexShard.prepareIndex(IndexShard.java:998)\n\tat org.elasticsearch.index.shard.IndexShard.applyIndexOperation(IndexShard.java:947)\n\tat org.elasticsearch.index.shard.IndexShard.applyIndexOperationOnPrimary(IndexShard.java:891)\n\tat org.elasticsearch.action.bulk.TransportShardBulkAction.executeBulkItemRequest(TransportShardBulkAction.java:320)\n\tat org.elasticsearch.action
```

### PR DESCRIPTION
**Pull request by weizijun**
Test case "tsdb/05_dimension_and_metric_in_non_tsdb_index/nested dimensions" failed is because just as @nik9000 said: 
Like ships passing in the night #83799 and #83837 missed each other and
then crashed together in the master branch causing failures.

The #83799 changes will cause `tsdb/05_dimension_and_metric_in_non_tsdb_index/nested` failed import from  #83837.
The reason is that the dimension added code is move from LuceneDocument to DocumentParserContext.
The nested type will create new LuceneDocument for each child, so the single value check will passed.
Now the check is in `DocumentDimensions`, and DocumentDimensions is a variable in DocumentParserContext. All fields of a doc use the same DocumentParserContext object, so the multi-nested-values doc will case the  multi-valued check to fail.

To pass the test, I remove the multi-nested-values.
But it may need a discuss the action of multi-nested-dimensions-values.

Here is the conclusion:
Reject `time_series_dimension` fields if there are any nested objects at all.

So this PR is change to implement the feature.

Closes: https://github.com/elastic/elasticsearch/issues/83915

### CODE REVIEW

**Comment by weizijun** [comment]
@nik9000 can you help to review?

**Comment by nik9000** [comment]
I just talked with a few folks and we'd talked about rejecting fields marked with `time_series_dimension` if there are any nested fields are all. We figured there probably is a use for them but we don't really know what it is yet. So, for now, we should disallow it and wait for someone to tell us how it *should* work.

Would you be interested in making that change?

**Comment by weizijun** [comment]
> Would you be interested in making that change?

of course yes. 
Is it the feature: reject the nested fields that are configured time_series_dimension?

**Comment by nik9000** [comment]
> : reject the nested fields that are configured time_series_dimension?

Reject `time_series_dimension` fields if there are *any* nested objects at all.

**Comment by weizijun** [comment]
> Reject `time_series_dimension` fields if there are _any_ nested objects at all.

got it.

**Comment by weizijun** [comment]
hi, @nik9000 , I have finished. Do you recommend implementing this feature in this PR, or opening another PR?

**Comment by nik9000** [comment]
> hi, @nik9000 , I have finished. Do you recommend implementing this feature in this PR, or opening another PR?

I'd open another PR and close this one.

**Comment by nik9000** [comment]
> I'd open another PR and close this one.

I see you already renamed everything in this PR. That's fine. I prefer a new one but it isn't worth doing now that you've done it the other way. I'll review and probably merge.

**Review by nik9000 — APPROVED** [review]
LGTM

@elasticmachine, ok to test

**Comment by nik9000** [comment]
@elasticmachine, test this please

**Comment by weizijun** [comment]
> I see you already renamed everything in this PR. That's fine. I prefer a new one but it isn't worth doing now that you've done it the other way. I'll review and probably merge.

Yeah, I implement it in this PR. It will fixed https://github.com/elastic/elasticsearch/issues/83915. Thanks @nik9000!

**Comment by nik9000** [comment]
@elasticmachine, test this please

**Review by nik9000 — COMMENTED** [review]
@elasticmachine test this please

**Comment by nik9000** [comment]
@elasticmachine, test this please

File: rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/tsdb/20_mapping.yml
Comment (weizijun) [inline rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/tsdb/20_mapping.yml:315]: this PR will change the nested failed message. the field dimension check is before tsdb nested check.

**Comment by nik9000** [comment]
@elasticmachine, test this please

**Comment by weizijun** [comment]
about the `org.elasticsearch.index.mapper.NestedObjectMapperTests.testNoDimensionNestedFields`, in -Dtests.iters=10000 tests, it's ok now.

**Comment by nik9000** [comment]
@elasticmachine test this please

**Comment by weizijun** [comment]
@nik9000 the failed bwc test is out of date, the previous bwc test is ok. I don't change the code that will affect the bwc test

**Comment by nik9000** [comment]
> This seems like an error.

@weizijun do you recall if you'd enabled "maintainer can push to this branch"? I think for you it's always true. We're just trying to figure out what's up with our tools.

**Comment by nik9000** [comment]
> I don't change it, but I see the flag is disabled. I will notice it next time.

Thanks! My *guess* is that it sets it to false once it's merged. We're going to try and set the bot to not comment on closed PRs. There really isn't any point regardless of what github's doing.

**Pull request merged**
